packages/list/index.js

```diff
@@ -0,0 +1,337 @@
+        const {listItemAttributes} = this.state;
+        attr = attr === 'tabindex' ? 'tabIndex' : attr;
+        const id = this.getListItemIdFromIndex_(index);
+        listItemAttributes[id][attr] = value;
+        this.setState({listItemAttributes});
+      },
+      removeAttributeForElementIndex: (index, attr) => {
+        const {listItemAttributes} = this.state;
+        attr = attr === 'tabindex' ? 'tabIndex' : attr;
+        const id = this.getListItemIdFromIndex_(index);
+        delete listItemAttributes[id][attr];
+        this.setState({listItemAttributes});
+      },
+      addClassForElementIndex: (index, className) => {
+        const {listItemClassList} = this.state;
+        const id = this.getListItemIdFromIndex_(index);
+        listItemClassList[id].add(className);
+        this.setState({listItemClassList});
+      },
+      removeClassForElementIndex: (index, className) => {
+        const {listItemClassList} = this.state;
+        const id = this.getListItemIdFromIndex_(index);
+        listItemClassList[id].delete(className);
+        this.setState({listItemClassList});
+      },
+      focusItemAtIndex: (index) => {
+        const id = this.getListItemIdFromIndex_(index);
+        const listItem = this.listItems_[id];
+        if (listItem) {
+          listItem.focus();
+        }
+      },
+      setTabIndexForListItemChildren: (listItemIndex, tabIndexValue) => {
+        const {listItemChildrenTabIndex} = this.state;
+        const id = this.getListItemIdFromIndex_(listItemIndex);
+        listItemChildrenTabIndex[id]= tabIndexValue;
+        this.setState({listItemChildrenTabIndex});
+      },
+      followHref: (index) => {
+        const id = this.getListItemIdFromIndex_(index);
+        const listItem = this.listItems_[id];
+        if (listItem) {
+          listItem.followHref();
+        }
+      },
+      toggleCheckbox: (index) => {
+        const id = this.getListItemIdFromIndex_(index);
+        const listItem = this.listItems_[id];
+        if (listItem) {
+          listItem.toggleCheckbox();
+        }
+      },
+    };
+  }
+
+  getListItemIdFromIndex_ = (index) => {
+    for (let id in this.listItemIndices_) {
+      if (this.listItemIndices_[id] === index) return id;
+    };
+    return '';
+  }
+
+  getIndexOfListItemElement_ = (element) => {
+    let listItemId = '';
+    for (let id in this.listItems_) {
+      if (this.listItems_[id].listItemElement === element) {
+        listItemId = id;
+        break;
+      }
+    }
+    return this.listItemIndices_[listItemId];
+  }
+
+  getListItemIndexOfTarget_ = (eventTarget) => {
+    let target = eventTarget;
+
+    // Find the first ancestor that is a list item.
+    while (this.getIndexOfListItemElement_(target) === undefined) {
+      if (!target || target === document) {
+        return -1;
+      }
+      target = target.parentElement;
+    }
+
+    return this.getIndexOfListItemElement_(target);
+  }
+
+  onKeyDown = (e) => {
```

Comment:
move onKeyDown, onClick, onFocus, and onBlur into the `renderListItem`, and pass the index. We can then remove `getListItemIndexOfTarget_`.